addon/services/router-scroll.js

```diff
@@ -29,6 +80,111 @@ class RouterScroll extends Service {
+  }
+
+  destroy() {
+    reset();
+
+    removeListener(this.router, 'routeWillChange', this._routeWillChange);
+    removeListener(this.router, 'routeDidChange', this._routeDidChange);
+
+    if (requestId) {
+      window.cancelAnimationFrame(requestId);
+    }
+
+    if (callbackRequestId) {
+      window.cancelAnimationFrame(callbackRequestId);
+    }
+
+    super.destroy(...arguments);
+  }
+
+  /**
+   * Updates the scroll position
+   * it will be a single transition
+   * @method updateScrollPosition
+   * @param {transition|transition[]} transition If before Ember 3.6, this will be an array of transitions, otherwise
+   */
+  updateScrollPosition(transition) {
+    const url = get(this, 'currentURL');
+    const hashElement = url ? document.getElementById(url.split('#').pop()) : null;
+
+    if (get(this, 'isFirstLoad')) {
+      this.unsetFirstLoad();
+    }
+
+    let scrollPosition;
+
+    if (url && url.indexOf('#') > -1 && hashElement) {
+      scrollPosition = { x: hashElement.offsetLeft, y: hashElement.offsetTop };
+    } else {
+      scrollPosition = get(this, 'position');
+    }
+
+    let preserveScrollPosition = (get(transition, 'router.currentRouteInfos') || []).some((routeInfo) => get(routeInfo, 'route.controller.preserveScrollPosition'));
+
+    // If `preserveScrollPosition` was not set on the controller, attempt fallback to `preserveScrollPosition` which was set on the router service.
+    if(!preserveScrollPosition) {
+      preserveScrollPosition = get(this, 'preserveScrollPosition')
+    }
+
+    if (!preserveScrollPosition) {
+      const scrollElement = get(this, 'scrollElement');
+      const targetElement = get(this, 'targetElement');
+
+      if (targetElement || 'window' === scrollElement) {
+        tryScrollRecursively(window.scrollTo, scrollPosition);
+      } else if ('#' === scrollElement.charAt(0)) {
+        const element = document.getElementById(scrollElement.substring(1));
+
+        if (element) {
+          let fn = (x, y) => {
+            element.scrollLeft = x;
+            element.scrollTop = y;
+          }
+          tryScrollRecursively(fn, scrollPosition, element);
+        }
+      }
+    }
+
+    sendEvent(this, 'didScroll', transition);
+  }
+
+  @action
+  _routeWillChange() {
+    if (get(this, 'isFastBoot')) {
+      return;
+    }
+
+    this.update();
+  }
+
+  @action
+  _routeDidChange(transition) {
+    if (get(this, 'isFastBoot')) {
+      return;
+    }
+
+    const scrollWhenIdle = get(this, 'scrollWhenIdle');
+    const scrollWhenAfterRender = get(this, 'scrollWhenAfterRender');
+
+    if (!scrollWhenIdle && !scrollWhenAfterRender) {
+      // out of the option, this happens on the tightest schedule
+      scheduleOnce('render', this, CALLBACK, transition);
+    } else if (scrollWhenAfterRender && !scrollWhenIdle) {
+      // out of the option, this happens on the tightest schedule
+      scheduleOnce('afterRender', this, CALLBACK, transition);
```

Comment:
Are the duplicated comments here correct, or is scheduling after `render` the option that happens on the tightest schedule?